staging/src/k8s.io/apiserver/pkg/server/hooks.go

```diff
@@ -135,6 +194,18 @@ func runPostStartHook(name string, entry postStartHookEntry, context PostStartHo
 	close(entry.done)
 }
 
+func runPreShutdownHook(name string, entry preShutdownHookEntry, context PreShutdownHookContext) {
+	var err error
+	func() {
+		// don't let the hook *accidentally* panic and kill the server
+		defer utilruntime.HandleCrash()
+		err = entry.hook(context)
+	}()
+	if err != nil {
+		glog.Warningf("PreShutdownHook %q failed: %v", name, err)
```

Comment:
we should collect the errors in the upper loop and return as an aggregated error, but keep calling the remaining hooks if one fails of course.

Reply:
This is tied into what we want to do with errors that I posted in the previous comment.  When is the advantage to returning all the errors vs just printing them here?  Is there some kind of control we wish to enable during shutdown if there is an error?

Reply:
Was thinking about tests which probably should fail, i.e. to notice problems during shutdown.

Reply:
why not just return the error? Seems like the most natural func signature. Why this bool construction?